File: pytet_v0.4/ctetris.py
from matrix import *
from tetris import *
from enum import Enum

class CTetris(Tetris):
    @classmethod
    def init(cls, setOfBlockArrays):
        Tetris.nBlockTypes = len(setOfBlockArrays)
        Tetris.nBlockDegrees = len(setOfBlockArrays[0])
        Tetris.setOfBlockObjects = [[0] * Tetris.nBlockDegrees for _ in range(Tetris.nBlockTypes)]
        colorblk = [[0] * Tetris.nBlockDegrees for _ in range(Tetris.nBlockTypes)]
        arrayBlk_maxSize = 0
        for i in range(Tetris.nBlockTypes):
            if arrayBlk_maxSize <= len(setOfBlockArrays[i][0]):
                arrayBlk_maxSize = len(setOfBlockArrays[i][0])
        Tetris.iScreenDw = arrayBlk_maxSize     # larget enough to cover the largest block
        for i in range(Tetris.nBlockTypes):
            for j in range(Tetris.nBlockDegrees):
                colorblk[i][j] = setOfBlockArrays[i][j]
                blkSize_x = len(setOfBlockArrays[i][j])
                for y in range(blkSize_x):
                    blkSize_y = len(setOfBlockArrays[i][j][y])
                    for x in range(blkSize_y):
                        if(colorblk[i][j][y][x] == 1):
                            colorblk[i][j][y][x] = i+1
                Tetris.setOfBlockObjects[i][j] = Matrix(colorblk[i][j])
                        
    def accept(self, key):
        self.state = TetrisState.Running

        if key >= '0' and key <= '6':
            if self.justStarted == False:
                self.deleteFullLines()
            self.iScreen = Matrix(self.oScreen)
            self.idxBlockType = int(key)
            self.idxBlockDegree = 0
            self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
            self.top = 0
            self.left = Tetris.iScreenDw + self.iScreenDx//2 - self.currBlk.get_dx()//2
            self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
            self.undoBlk = self.tempBlk.binary() + self.currBlk.binary() #undo judgement blk
            self.tempBlk = self.tempBlk + self.currBlk
            self.justStarted = False
            print()

            if self.undoBlk.anyGreaterThan(1):
                self.state = TetrisState.Finished
            self.oScreen = Matrix(self.iScreen)
            self.oScreen.paste(self.tempBlk, self.top, self.left)
            return self.state
        elif key == 'q':
            pass
        elif key == 'a': # move left
            self.left -= 1
        elif key == 'd': # move right
            self.left += 1
        elif key == 's': # move down
            self.top += 1
        elif key == 'w': # rotate the block clockwise
            self.idxBlockDegree = (self.idxBlockDegree + 1) % Tetris.nBlockDegrees
            self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
        elif key == ' ': # drop the block
            while not self.undoBlk.anyGreaterThan(1):
                    self.top += 1
                    self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
                    self.undoBlk = self.tempBlk.binary() + self.currBlk.binary() #undo judgement blk
                    self.tempBlk = self.tempBlk + self.currBlk
        else:
            print('Wrong key!!!')
            
        self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
        self.undoBlk = self.tempBlk.binary() + self.currBlk.binary() #undo judgement blk
        self.tempBlk = self.tempBlk + self.currBlk

        if self.undoBlk.anyGreaterThan(1):   ## 벽 충돌시 undo 수행
            if key == 'a': # undo: move right
                self.left += 1
            elif key == 'd': # undo: move left
                self.left -= 1
            elif key == 's': # undo: move up
                self.top -= 1
                self.state = TetrisState.NewBlock
            elif key == 'w': # undo: rotate the block counter-clockwise
                self.idxBlockDegree = (self.idxBlockDegree - 1) % Tetris.nBlockDegrees
                self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
            elif key == ' ': # undo: move up
                self.top -= 1
                self.state = TetrisState.NewBlock
            
            self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
            self.undoBlk = self.tempBlk.binary() + self.currBlk.binary()
            self.tempBlk = self.tempBlk + self.currBlk

        self.oScreen = Matrix(self.iScreen)
        self.oScreen.paste(self.tempBlk, self.top, self.left)

        return self.state
    
    def deleteFullLines(self):
        array = self.oScreen.get_array()
        deleteLine = False
        for y in range(self.oScreen.get_dy()-Tetris.iScreenDw):
            deleteLine = True
            for x in range(Tetris.iScreenDw, self.oScreen.get_dx()-Tetris.iScreenDw):
                if array[y][x] == 0:
                    deleteLine = False
                    break;
            if deleteLine == True:
                tempY = y
                for z in range(tempY, 0, -1):
                    for x in range(Tetris.iScreenDw, self.oScreen.get_dx()-Tetris.iScreenDw):
                        if z == 0:
                            array[z][x] = 0
                        else:
                            array[z][x] = array[z-1][x]
        self.oScreen = Matrix(array)
        return
    

File: cpptet_v0.5/CTetris.cpp
#include "CTetris.h"
CTetris::CTetris(int cy, int cx) : Tetris(cy, cx) {
    arrayScreen = createArrayScreen();
    iCScreen = Matrix(arrayScreen, arrayScreenDy, arrayScreenDx);
    oCScreen = Matrix(iCScreen);
}
int arrayBlks[7][4][4][4] =  {{{{0, 0, 1, 0},{0, 0, 1, 0},{0, 0, 1, 0},{0, 0, 1, 0}},
                               {{0, 0, 0, 0},{0, 0, 0, 0},{1, 1, 1, 1},{0, 0, 0, 0}},
                               {{0, 1, 0, 0},{0, 1, 0, 0},{0, 1, 0, 0},{0, 1, 0, 0}},
                               {{0, 0, 0, 0},{1, 1, 1, 1},{0, 0, 0, 0},{0, 0, 0, 0}}},
                              {{{1, 0, 0, 0},{1, 1, 1, 0},{0, 0, 0, 0},{0, 0, 0, 0}},
                               {{0, 1, 1, 0},{0, 1, 0, 0},{0, 1, 0, 0},{0, 0, 0, 0}},
                               {{0, 0, 0, 0},{1, 1, 1, 0},{0, 0, 1, 0},{0, 0, 0, 0}},
                               {{0, 1, 0, 0},{0, 1, 0, 0},{1, 1, 0, 0},{0, 0, 0, 0}}},
                              {{{0, 0, 1, 0},{1, 1, 1, 0},{0, 0, 0, 0},{0, 0, 0, 0}},
                               {{0, 1, 0, 0},{0, 1, 0, 0},{0, 1, 1, 0},{0, 0, 0, 0}},
                               {{0, 0, 0, 0},{1, 1, 1, 0},{1, 0, 0, 0},{0, 0, 0, 0}},
                               {{1, 1, 0, 0},{0, 1, 0, 0},{0, 1, 0, 0},{0, 0, 0, 0}}},
                              {{{0, 1, 1, 0},{0, 1, 1, 0},{0, 0, 0, 0},{0, 0, 0, 0}},
                               {{0, 1, 1, 0},{0, 1, 1, 0},{0, 0, 0, 0},{0, 0, 0, 0}},
                               {{0, 1, 1, 0},{0, 1, 1, 0},{0, 0, 0, 0},{0, 0, 0, 0}},
                               {{0, 1, 1, 0},{0, 1, 1, 0},{0, 0, 0, 0},{0, 0, 0, 0}}},
                              {{{0, 1, 1, 0},{1, 1, 0, 0},{0, 0, 0, 0},{0, 0, 0, 0}},
                               {{0, 1, 0, 0},{0, 1, 1, 0},{0, 0, 1, 0},{0, 0, 0, 0}},
                               {{0, 0, 0, 0},{0, 1, 1, 0},{1, 1, 0, 0},{0, 0, 0, 0}},
                               {{1, 0, 0, 0},{1, 1, 0, 0},{0, 1, 0, 0},{0, 0, 0, 0}}},
                              {{{0, 1, 0, 0},{1, 1, 1, 0},{0, 0, 0, 0},{0, 0, 0, 0}},
                               {{0, 1, 0, 0},{0, 1, 1, 0},{0, 1, 0, 0},{0, 0, 0, 0}},
                               {{0, 0, 0, 0},{1, 1, 1, 0},{0, 1, 0, 0},{0, 0, 0, 0}},
                               {{0, 1, 0, 0},{1, 1, 0, 0},{0, 1, 0, 0},{0, 0, 0, 0}}},
                              {{{1, 1, 0, 0},{0, 1, 1, 0},{0, 0, 0, 0},{0, 0, 0, 0}},
                               {{0, 0, 1, 0},{0, 1, 1, 0},{0, 1, 0, 0},{0, 0, 0, 0}},
                               {{0, 0, 0, 0},{1, 1, 0, 0},{0, 1, 1, 0},{0, 0, 0, 0}},
                               {{0, 1, 0, 0},{1, 1, 0, 0},{1, 0, 0, 0},{0, 0, 0, 0}}}};
CTetris::~CTetris(){
    for(int i = 0; i< nBlockTypes; i++)
        delete[] setOfCBlockObjects[i];
    delete[] setOfCBlockObjects;
}

Matrix** CTetris::setOfCBlockObjects =0;
void CTetris::init(int**** setOfBlockArrays, int max_type, int max_degree){
    setOfBlockArrays = new int***[max_type];
    for(int i=0;i<max_type;i++){
        setOfBlockArrays[i] = new int**[max_degree];
        for(int j=0;j<max_degree;j++){
            setOfBlockArrays[i][j] = new int*[4];
            for(int k=0;k<4;k++){
                setOfBlockArrays[i][j][k] = new int[4];
            }
        }
    }

    for(int i=0;i<max_type;i++){
        for(int j=0;j<max_degree;j++){
            for(int k=0;k<4;k++){
                for(int l=0;l<4;l++)
                setOfBlockArrays[i][j][k][l] = arrayBlks[i][j][k][l];
            }
        }
    }

    Tetris::init(setOfBlockArrays, max_type, max_degree);
    int arrayBlk_maxSize = 4;
    setOfCBlockObjects = new Matrix*[max_type];
    for(int i = 0;i<nBlockTypes;i++)
        setOfCBlockObjects[i] = new Matrix[nBlockDegrees];
    
    for(int i=0;i<nBlockTypes;i++){
        for(int j=0;j<nBlockDegrees;j++){
            int idxBlk[arrayBlk_maxSize * arrayBlk_maxSize];
            for(int k=0;k<arrayBlk_maxSize;k++)
                for(int l=0;l<arrayBlk_maxSize;l++)
                    idxBlk[k * arrayBlk_maxSize + l] = setOfBlockArrays[i][j][k][l];
            setOfCBlockObjects[i][j] = Matrix(idxBlk, arrayBlk_maxSize, arrayBlk_maxSize);
            setOfCBlockObjects[i][j].mulc(i+1);
        }
    }

    for(int i=0;i<max_type;i++){
        for(int j=0;j<max_degree;j++){
            for(int k=0;k<4;k++){
                delete[] setOfBlockArrays[i][j][k];
            }
            delete[] setOfBlockArrays[i][j];
        }
        delete[] setOfBlockArrays[i];
    }
    delete[] setOfBlockArrays;
}

TetrisState CTetris::accept(int key){
    if(key >= '0' && key <= '6'){
        if(justStarted == false)
            deleteFullLines();
        iCScreen = Matrix(oCScreen);
    }
    state = Tetris::accept(key);
    currCBlk = setOfCBlockObjects[idxBlockType][idxBlockDegree];
    tempBlk = iCScreen.clip(top, left, top+currCBlk.get_dy(),left+currCBlk.get_dx());
    tempBlk = tempBlk.add(&currCBlk);

    oCScreen = Matrix(iCScreen);
    oCScreen.paste(&tempBlk, top, left);

    return state;
}
void CTetris::deleteFullLines(){
    bool deleteline = false;
    int* c_array = new int[arrayScreenDy*arrayScreenDx];
    int** carray = oCScreen.get_array();
    for(int y=0;y<arrayScreenDy-iScreenDw;y++){
        deleteline = true;
        for(int x=iScreenDw;x<arrayScreenDx-iScreenDw;x++){
            if(carray[y][x] == 0){
                deleteline = false;
                break;
            }
        }
        if(deleteline == true){
            int tempY = y;
            for(int z=tempY;z>=0;z--){
                for(int x=iScreenDw;x<arrayScreenDx-iScreenDw;x++){
                    if(z == 0)
                        carray[z][x] = 0;
                    else
                        carray[z][x] = carray[z-1][x];
                }
            }

        }
    }
    for(int y =0;y<arrayScreenDy;y++)
        for(int x=0;x<arrayScreenDx;x++)
            c_array[y*arrayScreenDx + x] = carray[y][x];

    oCScreen = Matrix(c_array, arrayScreenDy, arrayScreenDx);

    delete[] c_array;
}